firmware/Core/Src/mpi/mpi_command_handling.c

```diff
@@ -55,65 +68,55 @@ uint8_t MPI_send_telecommand_get_response(const uint8_t *bytes_to_send, const si
+        // Timeout before receiving the first byte from the MPI
+        if (UART_mpi_buffer_write_idx == 0) {
+            if((HAL_GetTick() - UART_mpi_rx_start_time_ms) > MPI_RX_TIMEOUT_DURATION_MS) {
+                // Stop reception from the MPI & Reset mpi UART & transceiver mode states
+                MPI_set_transceiver_state(MPI_TRANSCEIVER_MODE_INACTIVE);
+                HAL_UART_DMAStop(UART_mpi_port_handle);
+                MPI_current_uart_rx_mode = MPI_RX_MODE_NOT_LISTENING_TO_MPI;
+                return 4; // Error code: Timeout waiting for 1st byte
+            }            
         }
 
-        // MPI has sent some data and has timed out
-        else if (
-            (*MPI_rx_buffer_len > 0)
-            && ((HAL_GetTick() - UART_mpi_rx_last_byte_write_time_ms) > MPI_RX_TIMEOUT_DURATION_MS)
-        ) {
-            // Stop reception from the MPI & Reset MPI UART mode state
-            MPI_set_transceiver_state(MPI_TRANSCEIVER_MODE_INACTIVE);
-            HAL_UART_DMAStop(&huart1);
-            MPI_current_uart_rx_mode = MPI_RX_MODE_NOT_LISTENING_TO_MPI;
-
-            // Copy the buffer to the last received byte index & clear the UART buffer
-            for (uint16_t i = 0; i < *MPI_rx_buffer_len; i++) {
-                MPI_rx_buffer[i] = UART_mpi_rx_buffer[i];
-            }
-            
-            // Clear the buffer (memset to 0, but volatile-compatible)
-            for (uint16_t i = 0; i < UART_mpi_rx_buffer_len; i++) {
-                UART_mpi_rx_buffer[i] = 0;
+        // Timeout in between (or end of) receiving bytes from the MPI
+        else {
+            const uint32_t current_time = HAL_GetTick();           // Get current time
+             if (
+                (current_time > UART_mpi_last_write_time_ms)       // Important seemingly-obvious safety check.
+                && ((current_time - UART_mpi_last_write_time_ms) > MPI_RX_TIMEOUT_DURATION_MS)
+            ) {
+                *MPI_rx_buffer_len = UART_mpi_buffer_write_idx;     // Set the length of the MPI response buffer
+                break;
             }
+        }
+    }
 
-            // Reset UART buffer write index
-            UART_mpi_rx_buffer_write_idx = 0;
+    // Stop reception from the MPI & Reset MPI transceiver mode state, Set MPI UART mode state to previous state
+    MPI_set_transceiver_state(MPI_TRANSCEIVER_MODE_INACTIVE);
+    HAL_UART_DMAStop(UART_mpi_port_handle);
+    MPI_current_uart_rx_mode = MPI_last_uart_rx_mode;
 
-            // Validate MPI response
-            const uint8_t MPI_response_result = MPI_validate_telecommand_response(bytes_to_send, MPI_rx_buffer, bytes_to_send_len);
+    // For now log the response from the MPI //TODO: Change this after testing
+    // Copy the buffer to the last received byte index & clear the UART buffer
+    for (uint16_t i = 0; i < *MPI_rx_buffer_len; i++) {
+        MPI_rx_buffer[i] = UART_mpi_buffer[i];
+        //UART_mpi_buffer[i] = 0;
```

Comment:
Although we clear the UART buffer before recording new responses, I was thinking we could clear them after they are read from here as well as part of a safe measure here. If it would be unnecessarily redundant I can remove it.